### DISCUSSION TITLE
Understanding shapes

### DISCUSSION THREAD

**Discussion by hmf** [deepjavalibrary/djl#755] [Q&A] [answered]
I am looking at chapter 5.2 of the online book. We have the following code:

```java
NDManager manager = NDManager.newBaseManager();

NDArray x = manager.randomUniform(0, 1, new Shape(2, 4));

Model model = Model.newInstance("lin-reg");

SequentialBlock net = new SequentialBlock();

net.add(Linear.builder().setUnits(8).build());
net.add(Activation.reluBlock());
net.add(Linear.builder().setUnits(1).build());
net.setInitializer(new NormalInitializer());
net.initialize(manager, DataType.FLOAT32, x.getShape());

model.setBlock(net);

Predictor<NDList, NDList> predictor = model.newPredictor(new NoopTranslator());

predictor.predict(new NDList(x)).singletonOrThrow(); // forward computation
```
I am assuming that given the `NoopTranslator` has no `Batchifier` the input may consist of one or more samples.  The input is `8` and the output `1`, so for an input of `Shape(1,8)` I should get a single value. 

However, looking at the example the input is `Shape(2,4)`. I don't see any `Blocks.batchFlattenBlock(8)`. In addition to this the output is `Shape(2,1)`. What transformations are occurring here? 

Apologies if this is simpleton's question.

TIA

**Comment by zachgk** [comment, marked as answer]
The input is not specified when creating the linear blocks in net. Because inputs can be computed, we thought it would be too much work to have to type them for every block. So, the 8 you are looking at is the **output** of the first linear block.

The input is not determined until the call to `net.initialize`. At this point, it determines the input is of size 4 (with batch size 2) using `x.getShape()`

**Reply by hmf** [reply]
Silly me. This was "explained" in chapter 5.1. Thank you.
